app/javascript/src/submit_handler.js

```diff
@@ -0,0 +1,108 @@
+
+const {
+  mergeObjects
+} = require('./utilities');
+
+class SubmitHandler {
+  #config;
+  #submitEnabled
+
+  constructor($form, config) {
+    this.#config = mergeObjects({
+      text: {
+        submitted: '',
+        unsubmitted: '',
+        submitting: '',
+      },
+      buttonSelector: ':submit',
+      buttonDescriptionSelector: '',
+      preventUnload: false,
+    }, config);
+    
+    this.$form = $form;
+    this.$button = this.$form.find(this.#config.buttonSelector);
+    this.text = this.#config.text;
+    if(this.#config.buttonDescriptionSelector) {
+      this.$buttonDescription = this.$form.find(this.#config.buttonDescriptionSelector);
+    }
+
+    this.$form.on('submit', (event) => this.#handleSubmit(event));
+    this.$button.on('click', (event) => this.#handleClick(event));
+  }
+
+  get required() {
+    return this.#submitEnabled;
+  }
+
+  set required(required) {
+    if(required) { 
+      this.#enable();
+    } else {
+      this.#disable();
+    }
+  }
+
+  forceSubmit() {
+    const currentState = this.#submitEnabled;
+    this.#submitEnabled = true;
+    this.$form.submit();
+    this.#submitEnabled = currentState;
+  }
+
+  #handleClick(event) {
+    if(this.required) {
+      this.#removeBeforeUnloadListener();
+      $(event.target).prop("value", this.#config.text.submitting );
```

Comment:
This is looking for the button, right? 
If so, we already have that with `this.$button` so assume we don't need to find it again with a lookup.

Also, just out of curiosity, is there a a different with `this.$button.val('blah')` and `this.$button.prop('value', 'blah')` that made you choose the latter?

One last thing (maybe it's the answer to previous 'curiosity' question)...

Does this work when `this.$button` is either:
a). `<input type="submit" value="I am an input submit" />`
b). `<button type="submit">I am a button submit</button>`

Reply:
as to `.prop('value', 'x')` vs `val('x')` there's no difference... not sure there was a reason why I chose one over the other.

As to your second point. That's a good one.  I'll handle that.